### PR TITLE
fix(ts-transformers): a plain-array callback's locals are body locals

### PR DESCRIPTION
**Pull request by Hixie**
A `map`, `filter`, or `find` callback on an ordinary JavaScript array runs eagerly while the pattern body is being built. Its statements are pattern body statements that happen to be written once and executed several times. The expression-site policy did not treat them that way: the `plain-array-value` callback boundary was excluded from `supportsPatternOwnedWrapperCallbackSite`, so a value binding inside such a callback was not a site the shared lowering model would lower into.

The consequence was a silently wrong render. In the fixed-column calendar shape

    {COLUMN_INDICES.map((colIdx) => {
      const isToday = weekDates?.[colIdx] === todayDate;
      return <div>{isToday ? "Today" : ""}</div>;
    })}

the comparison was emitted verbatim, so at run time it compared two reactive proxies by identity and produced a constant `false`. The "Today" column marker never appeared, and nothing said why. The same comparison lowers correctly at every neighboring position: at the top of the pattern body, inside an immediately-invoked arrow in JSX, inside a reactive `.map()` callback, and in this very callback once the ternary is part of the assigned expression rather than applied to a name bound earlier. Only the plainness of the mapped array turned it off, and that says nothing about whether the values the callback closes over are reactive.

The two diagnostics that would otherwise have caught it were both silent here. `isInRestrictedReactiveContext` exempts anything inside a JSX expression from the "wrap the computation in computed(...)" error, on the grounds that JSX is where lowering happens; the callback sits inside a JSX expression, so the exemption applied while the lowering did not. Written outside JSX the identical binding draws two errors.

The fix admits `plain-array-value` alongside `reactive-array-method`, `pattern-builder`, and `render-builder`. This is P-011 (pattern legality is lowerability) applied to a boundary that had been drawn by the receiver's type rather than by what the callback lowers to.

Two other behaviors follow from the same predicate, because the diagnostics and the lowering ask it the same question. A `.get()` read and an optional access inside a plain-array value callback now have a site to be carried by, so `["-", "+"].map((sep) => rows.get().join(sep))` compiles to an array of per-separator lifts instead of being rejected. A reactive array-method callback keeps the stricter rule, and the difference is structural rather than stylistic: that callback is lowered into a sub-pattern over per-element cells, so a read of the element binding has no pattern-body site to become a lift.

Every pattern in the tree was swept by instrumenting the policy and running `cfcheck` over all 420 files. Exactly one authored pattern was affected. imported-calendar carried a `computed(...)` wrapper on its `isToday`, with a comment saying the wrap existed because the name is read as a JSX condition further down rather than inside the expression it was written in. That is no longer a reason, so the wrapper and the comment both go. The pattern's own test pins the result: `assert_today_marked` counts the two places "Today" reaches the tree, and unwrapping without this transformer change drops that count to one. The sibling weekly-calendar wraps every per-column value the same way and keeps its wrappers, because there the wrap is uniform across the whole block rather than a note about one binding.

A second instrumented pass, looking for any reactive value computation still skipped at a pattern-context site, came back with five hits, all of them correct: three `??` fallbacks whose enclosing call carries them, and two event-handler attributes.

`packages/patterns/plain-array-callback-locals/` is a new lane-2 fixture covering the binding shapes side by side: the bare comparison, the same comparison with the ternary inside the binding, the indexed spelling, the `computed()` wrapper, a unary over the comparison, a comparison against an eager read, a falsy fallback, and a `.get()` in a second plain-array callback. Each probe renders a distinct string in both branches, because the shared vnode helpers' `textContent` maps a boolean to the empty string either way. Actions then move the comparison target and rewrite the array, so the test pins liveness rather than a single correct first render. Its contract baseline is recorded, which `deno task pattern-compat` requires of any pattern that yields one.

<a href="https://cubic.dev/pr/commontoolsinc/labs/pull/6016?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: packages/ts-transformers/src/policy/callback-boundary.ts [hunk 1]
```diff
@@ -404,6 +404,7 @@ export function getCallbackBoundarySemantics(
     isPatternToolCallback: supportedKind === "pattern-tool",
     supportsPatternOwnedWrapperCallbackSite: supportedKind ===
         "reactive-array-method" ||
+      supportedKind === "plain-array-value" ||
```
Comment (mathpirate) [inline packages/ts-transformers/src/policy/callback-boundary.ts]: [P1] Scope wrapper-site support to callbacks whose result may remain reactive

This is correct for the motivating plain-array `.map()` case, but `plain-array-value` is too broad an abstraction to admit wholesale. The flag is consumed by the shared expression-site policy, so it both makes callback-local expressions eligible for lift lowering and suppresses diagnostics that previously treated those expressions as unlowerable.

For `COLUMN_INDICES.map(...)`, that is sound: the callback executes during pattern construction, and each comparison can become a per-index lift capturing `weekDates`, `todayDate`, and `colIdx`. However, `classifyArrayCallbackContainerCall()` also returns `plain-array-value` for plain `filter`/`flatMap` calls and for non-void Array methods such as `find`. Those methods may synchronously interpret the callback result rather than merely collect it.

Two focused `cf check --show-transformed` probes demonstrate regressions introduced by this permission:

```ts
const filtered = VALUES.filter((value) => {
  const matches = value === target;
  return matches;
});
```

is emitted with `matches = __cfLift_2(...).for(...)`, followed by `return matches`. Native `filter()` therefore receives a reactive object, which is always truthy in JavaScript, and retains every element.

A named predicate inside `find()` is worse: the accepted source emitted a lift taking the callback-local `foundMatch` as an external input, referenced `target` without capturing it, and applied the lift with `foundMatch` outside its scope. The equivalent direct expression happened to lower correctly, but the newly supported local-binding form produces invalid output.

The focused upstream suite passes (46 tests / 624 steps), but the transformation fixture covers only `.map()`. The `.find()` policy test asserts only that this boolean is `true`; it does not inspect generated output. There is also a scope gap in the premise: `map`/`filter`/`flatMap` are initially recognized by method spelling without proving the receiver is an ordinary eager Array, and this boundary classification does not require the function to be callback argument zero.

Please split this boundary by method/callback role, or otherwise make this permission site-specific. Local wrapper sites are appropriate for a proven eager value-producing `.map()` callback. For `filter`, `find`, `some`, `every`, `sort`, `flatMap`, and similar callbacks whose result controls native execution, lower the owning operation as a whole or retain the diagnostic until that behavior is supported. Add transformed-output tests for named locals in `filter()` and `find()`, plus negative coverage for custom `.map()` methods and non-callback arguments.

**Review by mathpirate — COMMENTED** [review]
One boundary-scope correctness finding is included inline.

File: packages/ts-transformers/src/policy/callback-boundary.ts (hunk 1 above)
Comment (Hixie) [reply packages/ts-transformers/src/policy/callback-boundary.ts]: Claude responds:

Addressed in 0af644940.

## What changed

The permission no longer keys on the boundary kind. `supportsPatternOwnedWrapperCallbackSite` now gates `plain-array-value` on `isCollectingPlainArrayMethodCallback` (`ast/call-kind.ts`), which requires all three of:

- the callback is argument zero of its call;
- the call's **resolved signature** is declared on `Array` or `ReadonlyArray`;
- the method name is in `COLLECTING_ARRAY_METHOD_NAMES`, which contains `map` alone;

over a receiver that `classifyArrayMethodCallSite` reports as plain and not already lowered.

`map` stores each callback result without reading it, so a result that is a cell stays a cell. The other callback-taking array methods read the result while they run — `filter`/`find`/`some`/`every` as a boolean, `sort` as a number, `flatMap` as an array test, `reduce` as the next accumulator — and are excluded.

Both probes were reproduced before the change. The `filter` case emitted `const matches = __cfLift_2(...)` followed by `return matches`. The `find` case emitted a lift declaring the callback-local `foundMatch` as its own external input, with the comparison's other operand uncaptured, applied as `__cfLift_3({ foundMatch: foundMatch })` at pattern-body scope where no such binding exists. Both now emit the source expression unchanged and report `pattern-context:computation`.

## Tests added

Policy (`test/policy/callback-support.test.ts`), each asserting no wrapper site:

- `a plain filter callback carries no wrapper site`
- `a sort comparator carries no wrapper site`
- `a map on some other type carries no wrapper site`
- `a callback past argument zero carries no wrapper site`

`plain array find callbacks stay plain-array value callbacks` returns to asserting `false`. `plain array map callbacks stay plain-array value callbacks` now declares its own `interface Array<T>`, because the harness compiles with `noLib` and the resolved-owner test needs a signature to resolve.

Validation (`test/validation.test.ts`):

- `still errors on a named reactive comparison inside a plain filter`
- `still errors on a named reactive comparison inside a plain find`

`still errors on optional access inside a lowered array-method callback` is restored to its `main` form; this branch had converted it to an allowance.

## What was not changed

1. **The owning operation is not lowered as a whole.** Of the two options given, only the diagnostic retention is implemented. `filter`, `find`, `some`, `every`, `sort`, `flatMap`, and `reduce` callbacks keep the diagnostics they have on `main`; a reactive computation in one still requires `computed(...)`.

2. **No transformed-output goldens for `filter` and `find`.** The fixture runner fails any fixture whose transform reports an error, and both shapes now report one, so they cannot be golden pairs. The two validation tests above assert the diagnostic instead. The only transformed-output golden is `closures/map-plain-array-callback-local-comparison`, covering `map`.

3. **The two scope gaps are constrained on the new path only.** `classifyArrayMethodCall` still recognizes `map`/`filter`/`flatMap` by method spelling, and `classifyCallbackBoundary` still returns `plain-array-value` without requiring argument zero. Both are unchanged for every existing consumer. The resolved-owner and argument-zero tests live in `isCollectingPlainArrayMethodCallback`, so they bind the wrapper-site permission and nothing else.

## Gates

`deno task check`, `cfcheck` (420 patterns), `check-docs` (562 blocks), `pattern-compat` (337), `check-test-aliases`, `check-local-program`, `check-baselines-append-only`, `fmt`, `lint`; `packages/ts-transformers` 1148 tests; the `plain-array-callback-locals` and `imported-calendar` pattern tests. CI run 32212316710 completed successfully.

The target-language matrix, the current-behavior spec §6.5/§6.7, and design-deltas Addendum 4 were narrowed to state the `map`-only rule and to name the result-interpreting callbacks as unsupported.

File: packages/ts-transformers/src/ast/call-kind.ts [hunk 2]
```diff
@@ -835,6 +836,63 @@ export function classifyArrayCallbackContainerCall(
...
+ * each result in the output array without reading it, so a result that is a
+ * reactive cell stays a reactive cell.
+ *
+ * The other callback-taking Array methods interpret the result while they run:
+ * `filter`, `find`, `some`, and `every` read it as a boolean, `sort` reads it
+ * as a number, `flatMap` asks whether it is an array, and `reduce` feeds it
+ * back as the next accumulator. A cell reaching any of those is an object, so
+ * the method's own semantics change.
+ */
+const COLLECTING_ARRAY_METHOD_NAMES = new Set(["map"]);
```
Comment (mathpirate) [inline packages/ts-transformers/src/ast/call-kind.ts:850]: **[P1] Account for consumers of the mapped cells**

Checking only that `map` collects its callback result is insufficient: the resulting native array can immediately feed those cells into ordinary code that interprets them. For example:

```ts
const selected = ["a", "b"].map((value) => {
  const matches = value === target;
  return matches;
}).filter(Boolean);
```

The transform produces `[Cell(false), Cell(true)]`; native `filter(Boolean)` sees two truthy objects. A focused `cf test` expected one result and observed `subject.selected.length === 2`. The same problem reaches `find`, `some`, `every`, sorting, reductions, or a later consumer through a local binding. Restrict this permission to result flows that do not interpret the collected cells, or lower/reject the owning chain, and add a runtime pattern test rather than only a direct-JSX fixture.

File: packages/ts-transformers/src/ast/call-kind.ts [hunk 3]
```diff
@@ -835,6 +836,63 @@ export function classifyArrayCallbackContainerCall(
...
+ * resolved signature is declared on `Array`/`ReadonlyArray`, so a `map` of some
+ * other type does not qualify; and the receiver is a plain array that no
+ * reactive lowering owns, so the reactive collection operators keep their own
+ * structural treatment.
+ */
+export function isCollectingPlainArrayMethodCallback(
+  callback: ts.ArrowFunction | ts.FunctionExpression,
+  checker: ts.TypeChecker,
+): boolean {
+  const position = getCallArgumentPosition(callback);
```
Comment (mathpirate) [inline packages/ts-transformers/src/ast/call-kind.ts:870]: **[P1] Require a synchronous callback before granting wrapper sites**

This permission assumes the callback body executes during pattern construction, but async and generator callbacks violate that premise. This source currently passes validation:

```ts
["a", "b"].map(async (value) => {
  await Promise.resolve();
  const matches = value === target;
  return matches;
});
```

The generated lift runs after the `await`, outside the construction frame. A focused `cf test` logged two `SES_UNHANDLED_REJECTION` errors while still reporting the test and process as passing. A generator callback is admitted too even though `map` never executes its body. Exclude callbacks carrying `AsyncKeyword` or `asteriskToken` until those execution models have explicit lowering and coverage.

**Review by mathpirate — COMMENTED** [review]
Two post-merge P1 correctness findings from reviewing the final repaired head. Both remain reproducible on current main.
